sdk/core/Azure.Core/tests/ModelSerializationTests/Animal.cs

```diff
@@ -0,0 +1,159 @@
+            writer.WritePropertyName("isHungry"u8);
+            writer.WriteBooleanValue(IsHungry);
+            writer.WritePropertyName("weight"u8);
+            writer.WriteNumberValue(Weight);
+
+            if (options.HandleAdditionalProperties)
+            {
+                //write out the raw data
+                foreach (var property in RawData)
+                {
+                    writer.WritePropertyName(property.Key);
+#if NET6_0_OR_GREATER
+                    writer.WriteRawValue(property.Value);
+#else
+                    JsonSerializer.Serialize(writer, JsonDocument.Parse(property.Value.ToString()).RootElement);
+#endif
+                }
+            }
+            writer.WriteEndObject();
+        }
+
+        internal static Animal DeserializeAnimal(JsonElement element, SerializableOptions options)
+        {
+            double weight = default;
+            string name = "";
+            string latinName = "";
+            bool isHungry = default;
+
+            Dictionary<string, BinaryData> rawData = new Dictionary<string, BinaryData>();
+            foreach (var property in element.EnumerateObject())
+            {
+                if (property.NameEquals("weight"u8))
+                {
+                    weight = property.Value.GetDouble();
+                    continue;
+                }
+                if (property.NameEquals("name"u8))
+                {
+                    name = property.Value.GetString();
+                    continue;
+                }
+                if (property.NameEquals("latinName"u8))
+                {
+                    latinName = property.Value.GetString();
+                    continue;
+                }
+                if (property.NameEquals("isHungry"u8))
+                {
+                    isHungry = property.Value.GetBoolean();
+                    continue;
+                }
+
+                if (options.HandleAdditionalProperties)
+                {
+                    //this means it's an unknown property we got
+                    rawData.Add(property.Name, BinaryData.FromString(property.Value.GetRawText()));
+                }
+            }
+            return new Animal(weight, latinName, name, isHungry, rawData);
+        }
+        #endregion
+
+        #region InterfaceImplementation
+        public bool TryDeserialize(Stream stream, out long bytesConsumed, SerializableOptions options = default)
+        {
+            bytesConsumed = 0;
+            try
+            {
+                JsonDocument jsonDocument = JsonDocument.Parse(stream);
+                var model = DeserializeAnimal(jsonDocument.RootElement, options ?? new SerializableOptions());
+                this.LatinName = model.LatinName;
+                this.Weight = model.Weight;
+                this.IsHungry = model.IsHungry;
+                this.Name = model.Name;
+                this.RawData = model.RawData;
+                bytesConsumed = stream.Length;
```

Comment:
will `bytesConsumed` always be the same as `stream.Length` or would `stream.Position` be more accurate? 

If it is always `stream.Length`, do we add any value by returning this considering the caller already knows the length of the stream?

Reply:
same comment on Serialize

Reply:
For deserialize I think you are right position is probably more accurate, on serialize length and position should always be the same?

Reply:
But in both cases, are we adding any info that can't be gained by inspecting the `Stream`? What is the expected use case for `bytesWritten` and `bytesConsumed`?

Reply:
Thats a good question we should consider this when we finalize the signature sets this was part of the [original suggested strawman](https://github.com/Azure/autorest.csharp/issues/2158#issuecomment-1407139046) from @KrzysztofCwalina.